feature: add modal window filter expenses

ReddokDP opened the pull request:
feature: add modal window filter expenses

Limhill commented on src/modules/modals/component/ModalFilter.tsx:
```diff
@@ -0,0 +1,94 @@
+
+    // Сделаю валидацию в форме фильтра
+
+    return ReactDOM.createPortal(
```
не нужно создавать модальное окно через портал явно, ведь за тебя это уже сделано в ui-kit: https://admiralds.github.io/react-ui/current/?path=/docs/admiral-2-1-modal--docs

ReddokDP replied:
Fixed, теперь модальное окно отображается через условный рендеринг

Limhill commented on src/modules/modals/component/ModalFilter.tsx:
```diff
@@ -0,0 +1,94 @@
+    return ReactDOM.createPortal(
+        <ModalContentCustom
+            onClick={(e) => e.stopPropagation()}
+            onClose={() => dispatch(closeModalFilter())}
```
мы не пишем логику в jsx, помнишь? все функции в слушателях событий должны быть вынесены в хук.

ReddokDP replied:
Fixed, все логика вынесена в useModalFilter

Limhill commented on src/modules/modals/hooks/useModalFilter.ts:
```diff
@@ -0,0 +1,61 @@
+
+export const useModalFilter = () => {
+    const dispatch = useDispatch();
+    const { isOpen, modalData } = useSelector((state: RootState) => state.modal);
```
У таких селекторов будет проблема с производительностью: https://www.perplexity.ai/search/v-chem-problema-s-selektorami-HibVS5P6R0WTsY2G_HdzBQ

ReddokDP replied:
Fixed

Limhill commented on src/modules/modals/hooks/useModalFilter.ts:
```diff
@@ -0,0 +1,61 @@
+    const dispatch = useDispatch();
+    const { isOpen, modalData } = useSelector((state: RootState) => state.modal);
+
+    const { register, handleSubmit, reset } = useForm<ModalData>({
```
Кто писал этот код?)

В формах регистрации/логина ты оборачиваешь всё в Controller, а не используешь register и тут вдруг всё иначе)
Запиши себе в доку тогда, если вдруг не можешь запомнить как инпуты в форме привязывать к react-hook-form. Всегда в реальности делают через Controller: https://react-hook-form.com/docs/usecontroller/controller

ReddokDP replied:
Fixed, изначально неправильно обернул в Controller и соответственно связать логично не получилось, загуглил/нейронил способ через register. Исправил на Controller

Limhill commented on src/modules/modals/hooks/useModalFilter.ts:
```diff
@@ -0,0 +1,61 @@
+    });
+
+    const onSubmit: SubmitHandler<ModalData> = (data) => {
+        console.log(data);
```
console.log в проде не нужен

либо просматривай кажый файл перед коммитом, либо настрой линтер, чтобы тебе warning подчёркивал (жёлтеньким) на console.log-и

ReddokDP replied:
Fixed

Limhill commented on public/index.html:
```diff
@@ -1,11 +1,12 @@
+</head>
+<body>
+<div id="root"></div>
+<div id="modal-root"></div>
```
Это точно уйдёт, если использовать готовый компонент из ui-kit

ReddokDP replied:
Fixed

Limhill commented on src/modules/modals/slice/modalFilterSlice.ts:
```diff
@@ -0,0 +1,38 @@
+    modalData: null,
+};
+
+const modalFilterSlice = createSlice({
```
Как будто избыточно создавать целый слайс для управления одной модалкой.
Но значения фильтров передавать между компонентами надо, так что слайс всё-таки нужен. Просто это будет уже expenses slice, а не modal slice. 

Помнишь что мы выносим в слайс? Те вещи, которые не хотелось бы кидать пропсами через много уровней. Например данные формы фильтров. А вот состояние модалки можно было бы легко резолвить в локальном стейте в useFilterModal. Оттуда же возвращать пропсы для компонента modal или вообще целую модалку

ReddokDP replied:
Fixed, решили оставить целый слайс для управления, т.к. были проблемы с использованием локального стейта в кастомном хуке

Limhill commented on src/modules/expenses/components/Expenses.tsx:
```diff
@@ -20,11 +20,16 @@ export const Expenses = () => {
+                    <Button
+                        dimension="s"
+                        appearance="secondary"
+                        iconStart={SystemFilterOutline()}
```
Иконки лучше вставлять как компоненты, более явно получится.

```jsx
 <Button
    dimension="s"
    appearance="secondary"
    iconStart={<SystemFilterOutline />}
    onClick={handleOpenModal}>
    Фильтр
</Button>
```

ReddokDP replied:
fixed, везде вроде бы сменил

Limhill commented on src/modules/modals/component/ModalFilter.tsx:
```diff
@@ -0,0 +1,124 @@
+import { Button, InputField, Option, SelectField } from '@admiral-ds/react-ui';
+import {
+    ModalContentCustom,
+    ModalСontainerInputs,
```
Вот тут буква C - русская

ReddokDP replied:
fixed

Limhill commented on src/modules/modals/component/ModalFilter.tsx:
```diff
@@ -0,0 +1,124 @@
+                                placeholder="Введите актив"
+                                type="string"
+                                dimension="s"
+                                value={field.value ?? ''}
```
Инпуты бывают либо контролируемые, либо неконтролируемые.
Если контролируемые, то должны обязательно иметь value, onChange. Если неконтролируемые, то не должны иметь ни одного из этих пропсов и могут (необязательно) иметь defaultValue.
Короче это я к чему, тут value явно не надо прописывать или же надо прописывать onChange явно

ReddokDP replied:
fixed

Limhill commented on src/modules/modals/component/ModalFilter.tsx:
```diff
@@ -0,0 +1,124 @@
+                        control={control}
+                        rules={{ required: 'Начальная дата перевода обязательная!' }}
+                        render={({ field, fieldState }) => (
+                            <InputField
```
У тебя интересная конечно маска: mm/dd/yyyy, в РФ так не принято) Надо переделать на dd.MM.yyyy

ReddokDP replied:
![image](https://github.com/user-attachments/assets/b6f2d6cd-313a-414b-afd5-9d2f8efa8b30)
В форме и было dd.MM.yyyy
![image](https://github.com/user-attachments/assets/1d38faf0-64c8-446f-a674-9f5f4893ecb4)

Limhill commented on src/modules/modals/hooks/useModalFilter.ts:
```diff
@@ -0,0 +1,86 @@
+    value: string;
+}
+
+const selectorModalData = (state: RootState) => state.modal.modalData;
```
вытащи селекторы в слайс, чтобы всё в одном месте было

ReddokDP replied:
fixed, повытаскивал вообще во все селекторы в слайсы, которые уже были в проекте

Limhill commented on src/modules/modals/hooks/useModalFilter.ts:
```diff
@@ -0,0 +1,86 @@
+
+    const handleOpenModal = () => {
+        dispatch(openModalFilter());
+        if (modalData) {
```
Тут неочевидное пользовательское поведение. Если я написал что-то, не нажал "применить" и закрыл модалку, то я ожидаю, что форма очистится.

А сохраняться данные должны только если я нажал "применить"

ReddokDP replied:
fixed:
"Сбросить" - сбрасывает форму постоянно;
"Отмена"- не сохраняет изменения при заполненной форме и не сбрасывает изменения после нажатия;
"Принять" - запоминает поля формы.

Limhill commented:
1. Для карточек на homepage надо добавить cursor: pointer стиль и какое-то состояние disabled, иначе сейчас нифига непонятно какая карточка рабочая)
2. Модалки лучше в отдельный модуль не выносить, плохая идея. Это либо shared компонент, если там бизнес-логики нет. Либо компонент для конкретного бизнес-модуля, если бизнес-логика есть
3. shared/components/layout.tsx надо вынести в shared/layouts
4. все компоненты по всему проекту надо сделать следующим образом. Папка %componentName%, в ней Component.tsx и Components.styled.tsx. Так будет удобно.
5. routes перенести в app директорию
6. redux тоже перенести в app директорию
7. В половине проекта папки с компонентами называются component, в половине - components. Они ВСЕГДА должны называться во множественном числе. Сегодня там 1 компонент, а завтра 10 может быть, никто бегать и переименовывать не будет
8. В корне проекта лежит App.tsx, который ничего не делает и следовательно в проекте не нужен
9. modules/home/data/cardsData - это плохая история) Лучше сделать modules/home/constants - туда положить cardData.

ReddokDP commented:
> 1. Для карточек на homepage надо добавить cursor: pointer стиль и какое-то состояние disabled, иначе сейчас нифига непонятно какая карточка рабочая)
> 2. Модалки лучше в отдельный модуль не выносить, плохая идея. Это либо shared компонент, если там бизнес-логики нет. Либо компонент для конкретного бизнес-модуля, если бизнес-логика есть
> 3. shared/components/layout.tsx надо вынести в shared/layouts
> 4. все компоненты по всему проекту надо сделать следующим образом. Папка %componentName%, в ней Component.tsx и Components.styled.tsx. Так будет удобно.
> 5. routes перенести в app директорию
> 6. redux тоже перенести в app директорию
> 7. В половине проекта папки с компонентами называются component, в половине - components. Они ВСЕГДА должны называться во множественном числе. Сегодня там 1 компонент, а завтра 10 может быть, никто бегать и переименовывать не будет
> 8. В корне проекта лежит App.tsx, который ничего не делает и следовательно в проекте не нужен
> 9. modules/home/data/cardsData - это плохая история) Лучше сделать modules/home/constants - туда положить cardData.
-----
Выполнил все пункты кроме: 2,4. По ним есть парочку вопросов

Limhill commented:
0. Если пользователь залогинен, то при входе на корневой роут редиректить его на /homepage
1. При нажатии на "Сбросить фильтры", фильтры должны очищаться везде, в т.ч. в сторе. Для этого надо сделать отдельный редьюсер.
2. Задать локали для России, чтобы date-инпуты отображали российскую маску всегда
3. Переименовать modalSlice в expensesSlice и все интерфейсы/состояния аналогично (modal-что-то должны называться только те вещи, которые относятся конкретно к модалке)
4. Для имён инпутов в формах завести отдельный enum
5. В модуле auth объединить слайсы

ReddokDP commented:
> 0. Если пользователь залогинен, то при входе на корневой роут редиректить его на /homepage
> 1. При нажатии на "Сбросить фильтры", фильтры должны очищаться везде, в т.ч. в сторе. Для этого надо сделать отдельный редьюсер.
> 2. Задать локали для России, чтобы date-инпуты отображали российскую маску всегда
> 3. Переименовать modalSlice в expensesSlice и все интерфейсы/состояния аналогично (modal-что-то должны называться только те вещи, которые относятся конкретно к модалке)
> 4. Для имён инпутов в формах завести отдельный enum
> 5. В модуле auth объединить слайсы

----
0. fixed
1. fixed, но не делал отдельный редьюсер, обошелся без него с помощью ресета на изначальное значение и диспатча изначального значение в стор.
2. Просидел долгое время с этим моментом, мне не удалось решить. Сколько читал, перечитал, пронейронил - данная маска берется с параметров системы. И чтобы задать ее вручную, то будто адекватных способов нет и все костыльные через создание собственных функций, которые будут преобразовывать в нужную маску. По документации "Locales" в теме пробовал установить currentLocale: 'ru', но не помогло.
3. fixed. Вроде бы сделал, но чувствую что ошибся с некоторым переименованием
4. fixed. Сделаны 3 enum для формы авторизации, регистрации и фильтра затрат.
5. fixed.
